<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">

    <title>Java Source Code Analysis |  </title>
    <meta name="description" content="">
    <meta name="viewport" content="width=device-width, initial-scale=1">

    

    <!-- fonts -->
    <link href="//fonts.googleapis.com/css?family=Source+Sans+Pro:400,700" rel="stylesheet">
    <link href="//fonts.googleapis.com/css?family=Ubuntu:300,400,500,600,700" rel="stylesheet">

    <!-- stylesheets -->
    <link rel="stylesheet" href="/java-source-code-analysis/style/doc.css">

    <!-- favicon -->
    <link rel="icon" href="/java-source-code-analysis/favicon.ico">

    

  <meta name="generator" content="Hexo 6.2.0"></head>
  <body>

   <script>window.__INITIAL_STATE__ = {"page":{"path":"java/lang/Throwable.html","title":""},"data":{"navigation":{"logo":{"text":"Java 源码分析","type":"link","path":"index.html"}}},"config":{"timezone":"Asia/Shanghai","root":"/java-source-code-analysis/","time_format":"HH:mm:ss","theme":"doc","theme_config":{"swagger_ui":{"version":2,"permalinks":true,"api_explorer":true,"download":"Download specification","show_extensions":false,"deep_linking":true,"display_operation_id":false,"doc_expansion":"none"},"search":{"skip":false,"background":false,"route":"/lunr.json"},"favicon":"/favicon.ico"}}}</script>

    <div id="react-navigation-root"><div class="doc-navigation" data-reactroot=""><nav class="doc-navbar"><a href="/java-source-code-analysis/index.html" class="doc-navbar__logo"><img src="/java-source-code-analysis/images/logo.png" class="doc-navbar__logo__img"/><span class="doc-navbar__logo__text">Java 源码分析</span></a><i class="dc-icon dc-icon--close dc-icon--interactive doc-sidebar-close doc-navbar__sidebar-close doc-navbar__sidebar-close--desktop"></i><i class="dc-icon dc-icon--menu dc-icon--interactive doc-sidebar-toggle doc-navbar__sidebar-toggle"></i></nav><nav class="doc-sidebar"><div class="doc-sidebar__vertical-menu"><i class="dc-icon dc-icon--menu dc-icon--interactive doc-sidebar-toggle doc-sidebar-toggle--primary doc-sidebar__vertical-menu__item"></i><i class="dc-icon dc-icon--search dc-icon--interactive doc-sidebar__vertical-menu__item doc-sidebar__vertical-menu__item--primary"></i></div><div class="doc-sidebar-content"><div class="doc-sidebar__search-form"></div><ul class="doc-sidebar-list"></ul></div></nav></div></div>
    <div class="doc-content">
  <div class="dc-page">
    <div class="dc-card">
      <div id="react-search-results-root"></div>
      <div id="page-content" class="doc-formatting">
        <h2 id="Throwable"><a href="#Throwable" class="headerlink" title="Throwable"></a>Throwable</h2><table>
<thead>
<tr>
<th align="center">类名</th>
<th align="center">类型</th>
<th align="center">父接口</th>
<th align="center">权限</th>
<th align="center">static</th>
<th align="center">final</th>
<th align="center">描述</th>
</tr>
</thead>
<tbody><tr>
<td align="center"><strong>java.lang.Throwable</strong></td>
<td align="center">class</td>
<td align="center">Serializable</td>
<td align="center">public</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center"></td>
</tr>
</tbody></table>
<h2 id=""><a href="#" class="headerlink" title=""></a></h2><table>
<thead>
<tr>
<th align="center">参数</th>
<th align="center">默认值</th>
<th align="center">权限</th>
<th align="center">static</th>
<th align="center">final</th>
<th align="center">volatile</th>
<th align="center">synchronized</th>
<th align="center">transient</th>
<th align="left">描述</th>
</tr>
</thead>
<tbody><tr>
<td align="center"><strong>serialVersionUID: long</strong></td>
<td align="center">-3042686055658047285L</td>
<td align="center">private</td>
<td align="center"><strong><font color='red'>是</font></strong></td>
<td align="center"><strong><font color='red'>是</font></strong></td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">use serialVersionUID from JDK 1.0.2 for interoperability</td>
</tr>
<tr>
<td align="center"><strong>backtrace: Object</strong></td>
<td align="center">null</td>
<td align="center">private</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center"><strong><font color='red'>是</font></strong></td>
<td align="left">Native code saves some indication of the stack backtrace in this slot.</td>
</tr>
<tr>
<td align="center"><strong>detailMessage: String</strong></td>
<td align="center">null</td>
<td align="center">private</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">Specific details about the Throwable.  For example, for {@code FileNotFoundException}, this contains the name of the file that could not be found.  @serial</td>
</tr>
<tr>
<td align="center"><strong>UNASSIGNED_STACK: StackTraceElement[]</strong></td>
<td align="center">new StackTraceElement[0]</td>
<td align="center">private</td>
<td align="center"><strong><font color='red'>是</font></strong></td>
<td align="center"><strong><font color='red'>是</font></strong></td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">A shared value for an empty stack.</td>
</tr>
<tr>
<td align="center"><strong>cause: Throwable</strong></td>
<td align="center">this</td>
<td align="center">private</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">The throwable that caused this throwable to get thrown, or null if this throwable was not caused by another throwable, or if the causative throwable is unknown.  If this field is equal to this throwable itself, it indicates that the cause of this throwable has not yet been initialized.  @serial @since 1.4</td>
</tr>
<tr>
<td align="center"><strong>stackTrace: StackTraceElement[]</strong></td>
<td align="center">UNASSIGNED_STACK</td>
<td align="center">private</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">The stack trace, as returned by {@link #getStackTrace()}.  The field is initialized to a zero-length array.  A {@code null} value of this field indicates subsequent calls to {@link #setStackTrace(StackTraceElement[])} and {@link #fillInStackTrace()} will be be no-ops.  @serial @since 1.4</td>
</tr>
<tr>
<td align="center"><strong>SUPPRESSED_SENTINEL: List&lt;Throwable&gt;</strong></td>
<td align="center">Collections.unmodifiableList(new ArrayList<Throwable>(0))</td>
<td align="center">private</td>
<td align="center"><strong><font color='red'>是</font></strong></td>
<td align="center"><strong><font color='red'>是</font></strong></td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left"></td>
</tr>
<tr>
<td align="center"><strong>suppressedExceptions: List&lt;Throwable&gt;</strong></td>
<td align="center">SUPPRESSED_SENTINEL</td>
<td align="center">private</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">The list of suppressed exceptions, as returned by {@link #getSuppressed()}.  The list is initialized to a zero-element unmodifiable sentinel list.  When a serialized Throwable is read in, if the {@code suppressedExceptions} field points to a zero-element list, the field is reset to the sentinel value.  @serial @since 1.7</td>
</tr>
<tr>
<td align="center"><strong>NULL_CAUSE_MESSAGE: String</strong></td>
<td align="center">“Cannot suppress a null exception.”</td>
<td align="center">private</td>
<td align="center"><strong><font color='red'>是</font></strong></td>
<td align="center"><strong><font color='red'>是</font></strong></td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">Message for trying to suppress a null exception.</td>
</tr>
<tr>
<td align="center"><strong>SELF_SUPPRESSION_MESSAGE: String</strong></td>
<td align="center">“Self-suppression not permitted”</td>
<td align="center">private</td>
<td align="center"><strong><font color='red'>是</font></strong></td>
<td align="center"><strong><font color='red'>是</font></strong></td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">Message for trying to suppress oneself.</td>
</tr>
<tr>
<td align="center"><strong>CAUSE_CAPTION: String</strong></td>
<td align="center">“Caused by: “</td>
<td align="center">private</td>
<td align="center"><strong><font color='red'>是</font></strong></td>
<td align="center"><strong><font color='red'>是</font></strong></td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">Caption  for labeling causative exception stack traces</td>
</tr>
<tr>
<td align="center"><strong>SUPPRESSED_CAPTION: String</strong></td>
<td align="center">“Suppressed: “</td>
<td align="center">private</td>
<td align="center"><strong><font color='red'>是</font></strong></td>
<td align="center"><strong><font color='red'>是</font></strong></td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">Caption for labeling suppressed exception stack traces</td>
</tr>
<tr>
<td align="center"><strong>EMPTY_THROWABLE_ARRAY: Throwable[]</strong></td>
<td align="center">new Throwable[0]</td>
<td align="center">private</td>
<td align="center"><strong><font color='red'>是</font></strong></td>
<td align="center"><strong><font color='red'>是</font></strong></td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left"></td>
</tr>
</tbody></table>
<h2 id="-1"><a href="#-1" class="headerlink" title=""></a></h2><table>
<thead>
<tr>
<th align="center">方法</th>
<th align="center">权限</th>
<th align="center">static</th>
<th align="center">final</th>
<th align="center">volatile</th>
<th align="center">synchronized</th>
<th align="center">transient</th>
<th align="center">abstract</th>
<th align="left">描述</th>
<th align="left">异常</th>
</tr>
</thead>
<tbody><tr>
<td align="center"><strong>Throwable()</strong></td>
<td align="center">public</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">Constructs a new throwable with ‘null’ as its detail message. The cause is not initialized, and may subsequently be initialized by a call to ‘#initCause’.  <p>The ‘#fillInStackTrace()’ method is called to initialize the stack trace data in the newly created throwable.</td>
<td align="left"></td>
</tr>
<tr>
<td align="center"><strong>Throwable(String message)</strong></td>
<td align="center">public</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">Constructs a new throwable with the specified detail message.  The cause is not initialized, and may subsequently be initialized by a call to ‘#initCause’.  <p>The ‘#fillInStackTrace()’ method is called to initialize the stack trace data in the newly created throwable.</td>
<td align="left"></td>
</tr>
<tr>
<td align="center"><strong>Throwable(String message, Throwable cause)</strong></td>
<td align="center">public</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">Constructs a new throwable with the specified detail message and cause.  <p>Note that the detail message associated with ‘cause’ is <i>not</i> automatically incorporated in this throwable’s detail message.  <p>The ‘#fillInStackTrace()’ method is called to initialize the stack trace data in the newly created throwable.</td>
<td align="left"></td>
</tr>
<tr>
<td align="center"><strong>Throwable(Throwable cause)</strong></td>
<td align="center">public</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">Constructs a new throwable with the specified cause and a detail message of ‘(cause&#x3D;&#x3D;null ? null : cause.toString())’ (which typically contains the class and detail message of ‘cause’). This constructor is useful for throwables that are little more than wrappers for other throwables (for example, ‘java.security.PrivilegedActionException’).  <p>The ‘#fillInStackTrace()’ method is called to initialize the stack trace data in the newly created throwable.</td>
<td align="left"></td>
</tr>
<tr>
<td align="center"><strong>Throwable(String message, Throwable cause, boolean enableSuppression, boolean writableStackTrace)</strong></td>
<td align="center">protected</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">Constructs a new throwable with the specified detail message, cause, ‘#addSuppressed suppression’ enabled or disabled, and writable stack trace enabled or disabled.  If suppression is disabled, ‘#getSuppressed’ for this object will return a zero-length array and calls to ‘#addSuppressed’ that would otherwise append an exception to the suppressed list will have no effect.  If the writable stack trace is false, this constructor will not call ‘#fillInStackTrace()’, a ‘null’ will be written to the ‘stackTrace’ field, and subsequent calls to ‘fillInStackTrace’ and ‘#setStackTrace(StackTraceElement[])’ will not set the stack trace.  If the writable stack trace is false, ‘#getStackTrace’ will return a zero length array.  <p>Note that the other constructors of ‘Throwable’ treat suppression as being enabled and the stack trace as being writable.  Subclasses of ‘Throwable’ should document any conditions under which suppression is disabled and document conditions under which the stack trace is not writable. Disabling of suppression should only occur in exceptional circumstances where special requirements exist, such as a virtual machine reusing exception objects under low-memory situations.  Circumstances where a given exception object is repeatedly caught and rethrown, such as to implement control flow between two sub-systems, is another situation where immutable throwable objects would be appropriate.</td>
<td align="left"></td>
</tr>
<tr>
<td align="center"><strong>getMessage(): String</strong></td>
<td align="center">public</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">Returns the detail message string of this throwable.</td>
<td align="left"></td>
</tr>
<tr>
<td align="center"><strong>getLocalizedMessage(): String</strong></td>
<td align="center">public</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">Creates a localized description of this throwable. Subclasses may override this method in order to produce a locale-specific message.  For subclasses that do not override this method, the default implementation returns the same result as ‘getMessage()’.</td>
<td align="left"></td>
</tr>
<tr>
<td align="center"><strong>getCause(): Throwable</strong></td>
<td align="center">public</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center"><strong><font color='red'>是</font></strong></td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">Returns the cause of this throwable or ‘null’ if the cause is nonexistent or unknown.  (The cause is the throwable that caused this throwable to get thrown.)  <p>This implementation returns the cause that was supplied via one of the constructors requiring a ‘Throwable’, or that was set after creation with the ‘#initCause(Throwable)’ method.  While it is typically unnecessary to override this method, a subclass can override it to return a cause set by some other means.  This is appropriate for a “legacy chained throwable” that predates the addition of chained exceptions to ‘Throwable’.  Note that it is <i>not</i> necessary to override any of the ‘PrintStackTrace’ methods, all of which invoke the ‘getCause’ method to determine the cause of a throwable.</td>
<td align="left"></td>
</tr>
<tr>
<td align="center"><strong>initCause(Throwable cause): Throwable</strong></td>
<td align="center">public</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center"><strong><font color='red'>是</font></strong></td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">Initializes the <i>cause</i> of this throwable to the specified value. (The cause is the throwable that caused this throwable to get thrown.)  <p>This method can be called at most once.  It is generally called from within the constructor, or immediately after creating the throwable.  If this throwable was created with ‘#Throwable(Throwable)’ or ‘#Throwable(String,Throwable)’, this method cannot be called even once.  <p>An example of using this method on a legacy throwable type without other support for setting the cause is:  <pre> try { lowLevelOp(); } catch (LowLevelException le) { throw (HighLevelException) new HighLevelException().initCause(le); } </pre></td>
<td align="left">IllegalStateException<br/>IllegalArgumentException</td>
</tr>
<tr>
<td align="center"><strong>toString(): String</strong></td>
<td align="center">public</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">Returns a short description of this throwable. The result is the concatenation of: <ul> <li> the ‘Class#getName() name’ of the class of this object <li> “: “ (a colon and a space) <li> the result of invoking this object’s ‘#getLocalizedMessage’ method </ul> If ‘getLocalizedMessage’ returns ‘null’, then just the class name is returned.</td>
<td align="left"></td>
</tr>
<tr>
<td align="center"><strong>printStackTrace(): void</strong></td>
<td align="center">public</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">Prints this throwable and its backtrace to the standard error stream. This method prints a stack trace for this ‘Throwable’ object on the error output stream that is the value of the field ‘System.err’. The first line of output contains the result of the ‘#toString()’ method for this object.  Remaining lines represent data previously recorded by the method ‘#fillInStackTrace()’. The format of this information depends on the implementation, but the following example may be regarded as typical: <blockquote><pre> java.lang.NullPointerException at MyClass.mash(MyClass.java:9) at MyClass.crunch(MyClass.java:6) at MyClass.main(MyClass.java:3) </pre></blockquote> This example was produced by running the program: <pre> class MyClass { public static void main(String[] args) { crunch(null); } static void crunch(int[] a) { mash(a); } static void mash(int[] b) { System.out.println(b[0]); } } </pre> The backtrace for a throwable with an initialized, non-null cause should generally include the backtrace for the cause.  The format of this information depends on the implementation, but the following example may be regarded as typical: <pre> HighLevelException: MidLevelException: LowLevelException at Junk.a(Junk.java:13) at Junk.main(Junk.java:4) Caused by: MidLevelException: LowLevelException at Junk.c(Junk.java:23) at Junk.b(Junk.java:17) at Junk.a(Junk.java:11) ... 1 more Caused by: LowLevelException at Junk.e(Junk.java:30) at Junk.d(Junk.java:27) at Junk.c(Junk.java:21) ... 3 more </pre> Note the presence of lines containing the characters ‘“…”‘. These lines indicate that the remainder of the stack trace for this exception matches the indicated number of frames from the bottom of the stack trace of the exception that was caused by this exception (the “enclosing” exception).  This shorthand can greatly reduce the length of the output in the common case where a wrapped exception is thrown from same method as the “causative exception” is caught.  The above example was produced by running the program: <pre> public class Junk { public static void main(String args[]) { try { a(); } catch(HighLevelException e) { e.printStackTrace(); } } static void a() throws HighLevelException { try { b(); } catch(MidLevelException e) { throw new HighLevelException(e); } } static void b() throws MidLevelException { c(); } static void c() throws MidLevelException { try { d(); } catch(LowLevelException e) { throw new MidLevelException(e); } } static void d() throws LowLevelException { e(); } static void e() throws LowLevelException { throw new LowLevelException(); } }  class HighLevelException extends Exception { HighLevelException(Throwable cause) { super(cause); } }  class MidLevelException extends Exception { MidLevelException(Throwable cause)  { super(cause); } }  class LowLevelException extends Exception { } </pre> As of release 7, the platform supports the notion of <i>suppressed exceptions</i> (in conjunction with the ‘try’-with-resources statement). Any exceptions that were suppressed in order to deliver an exception are printed out beneath the stack trace.  The format of this information depends on the implementation, but the following example may be regarded as typical:  <pre> Exception in thread "main" java.lang.Exception: Something happened at Foo.bar(Foo.java:10) at Foo.main(Foo.java:5) Suppressed: Resource$CloseFailException: Resource ID = 0 at Resource.close(Resource.java:26) at Foo.bar(Foo.java:9) ... 1 more </pre> Note that the “… n more” notation is used on suppressed exceptions just at it is used on causes. Unlike causes, suppressed exceptions are indented beyond their “containing exceptions.”  <p>An exception can have both a cause and one or more suppressed exceptions: <pre> Exception in thread "main" java.lang.Exception: Main block at Foo3.main(Foo3.java:7) Suppressed: Resource$CloseFailException: Resource ID = 2 at Resource.close(Resource.java:26) at Foo3.main(Foo3.java:5) Suppressed: Resource$CloseFailException: Resource ID = 1 at Resource.close(Resource.java:26) at Foo3.main(Foo3.java:5) Caused by: java.lang.Exception: I did it at Foo3.main(Foo3.java:8) </pre> Likewise, a suppressed exception can have a cause: <pre> Exception in thread "main" java.lang.Exception: Main block at Foo4.main(Foo4.java:6) Suppressed: Resource2$CloseFailException: Resource ID = 1 at Resource2.close(Resource2.java:20) at Foo4.main(Foo4.java:5) Caused by: java.lang.Exception: Rats, you caught me at Resource2$CloseFailException.&lt;init&gt;(Resource2.java:45) ... 2 more </pre></td>
<td align="left"></td>
</tr>
<tr>
<td align="center"><strong>printStackTrace(PrintStream s): void</strong></td>
<td align="center">public</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">Prints this throwable and its backtrace to the specified print stream.</td>
<td align="left"></td>
</tr>
<tr>
<td align="center"><strong>printStackTrace(PrintStreamOrWriter s): void</strong></td>
<td align="center">private</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr>
<td align="center"><strong>printEnclosedStackTrace(PrintStreamOrWriter s, StackTraceElement[] enclosingTrace, String caption, String prefix, Set&lt;Throwable&gt; dejaVu): void</strong></td>
<td align="center">private</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">Print our stack trace as an enclosed exception for the specified stack trace.</td>
<td align="left"></td>
</tr>
<tr>
<td align="center"><strong>printStackTrace(PrintWriter s): void</strong></td>
<td align="center">public</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">Prints this throwable and its backtrace to the specified print writer.</td>
<td align="left"></td>
</tr>
<tr>
<td align="center"><strong>fillInStackTrace(): Throwable</strong></td>
<td align="center">public</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center"><strong><font color='red'>是</font></strong></td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">Fills in the execution stack trace. This method records within this ‘Throwable’ object information about the current state of the stack frames for the current thread.  <p>If the stack trace of this ‘Throwable’ ‘Throwable#Throwable(String, Throwable, boolean, boolean) is not writable’, calling this method has no effect.</td>
<td align="left"></td>
</tr>
<tr>
<td align="center"><strong>fillInStackTrace(int dummy): Throwable</strong></td>
<td align="center">private</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr>
<td align="center"><strong>getStackTrace(): StackTraceElement[]</strong></td>
<td align="center">public</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">Provides programmatic access to the stack trace information printed by ‘#printStackTrace()’.  Returns an array of stack trace elements, each representing one stack frame.  The zeroth element of the array (assuming the array’s length is non-zero) represents the top of the stack, which is the last method invocation in the sequence.  Typically, this is the point at which this throwable was created and thrown. The last element of the array (assuming the array’s length is non-zero) represents the bottom of the stack, which is the first method invocation in the sequence.  <p>Some virtual machines may, under some circumstances, omit one or more stack frames from the stack trace.  In the extreme case, a virtual machine that has no stack trace information concerning this throwable is permitted to return a zero-length array from this method.  Generally speaking, the array returned by this method will contain one element for every frame that would be printed by ‘printStackTrace’.  Writes to the returned array do not affect future calls to this method.</td>
<td align="left"></td>
</tr>
<tr>
<td align="center"><strong>getOurStackTrace(): StackTraceElement[]</strong></td>
<td align="center">private</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center"><strong><font color='red'>是</font></strong></td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left"></td>
<td align="left"></td>
</tr>
<tr>
<td align="center"><strong>setStackTrace(StackTraceElement[] stackTrace): void</strong></td>
<td align="center">public</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">Sets the stack trace elements that will be returned by ‘#getStackTrace()’ and printed by ‘#printStackTrace()’ and related methods.  This method, which is designed for use by RPC frameworks and other advanced systems, allows the client to override the default stack trace that is either generated by ‘#fillInStackTrace()’ when a throwable is constructed or deserialized when a throwable is read from a serialization stream.  <p>If the stack trace of this ‘Throwable’ ‘Throwable#Throwable(String, Throwable, boolean, boolean) is not writable’, calling this method has no effect other than validating its argument.</td>
<td align="left">NullPointerException</td>
</tr>
<tr>
<td align="center"><strong>getStackTraceDepth(): int</strong></td>
<td align="center">default</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">Returns the number of elements in the stack trace (or 0 if the stack trace is unavailable).  package-protection for use by SharedSecrets.</td>
<td align="left"></td>
</tr>
<tr>
<td align="center"><strong>getStackTraceElement(int index): StackTraceElement</strong></td>
<td align="center">default</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">Returns the specified element of the stack trace.  package-protection for use by SharedSecrets.</td>
<td align="left">IndexOutOfBoundsException</td>
</tr>
<tr>
<td align="center"><strong>readObject(ObjectInputStream s): void</strong></td>
<td align="center">private</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">Reads a ‘Throwable’ from a stream, enforcing well-formedness constraints on fields.  Null entries and self-pointers are not allowed in the list of ‘suppressedExceptions’.  Null entries are not allowed for stack trace elements.  A null stack trace in the serial form results in a zero-length stack element array. A single-element stack trace whose entry is equal to ‘new StackTraceElement(“”, “”, null, Integer.MIN_VALUE)’ results in a ‘null’ ‘stackTrace’ field.  Note that there are no constraints on the value the ‘cause’ field can hold; both ‘null’ and ‘this’ are valid values for the field.</td>
<td align="left">ClassNotFoundException<br/>IOException</td>
</tr>
<tr>
<td align="center"><strong>writeObject(ObjectOutputStream s): void</strong></td>
<td align="center">private</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center">否</td>
<td align="center"><strong><font color='red'>是</font></strong></td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">Write a ‘Throwable’ object to a stream.  A ‘null’ stack trace field is represented in the serial form as a one-element array whose element is equal to ‘new StackTraceElement(“”, “”, null, Integer.MIN_VALUE)’.</td>
<td align="left">IOException</td>
</tr>
<tr>
<td align="center"><strong>addSuppressed(Throwable exception): void</strong></td>
<td align="center">public</td>
<td align="center">否</td>
<td align="center"><strong><font color='red'>是</font></strong></td>
<td align="center">否</td>
<td align="center"><strong><font color='red'>是</font></strong></td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">Appends the specified exception to the exceptions that were suppressed in order to deliver this exception. This method is thread-safe and typically called (automatically and implicitly) by the ‘try’-with-resources statement.  <p>The suppression behavior is enabled <em>unless</em> disabled ‘#Throwable(String, Throwable, boolean, boolean) via a constructor’.  When suppression is disabled, this method does nothing other than to validate its argument.  <p>Note that when one exception ‘#initCause(Throwable) causes’ another exception, the first exception is usually caught and then the second exception is thrown in response.  In other words, there is a causal connection between the two exceptions.  In contrast, there are situations where two independent exceptions can be thrown in sibling code blocks, in particular in the ‘try’ block of a ‘try’-with-resources statement and the compiler-generated ‘finally’ block which closes the resource.  In these situations, only one of the thrown exceptions can be propagated.  In the ‘try’-with-resources statement, when there are two such exceptions, the exception originating from the ‘try’ block is propagated and the exception from the ‘finally’ block is added to the list of exceptions suppressed by the exception from the ‘try’ block.  As an exception unwinds the stack, it can accumulate multiple suppressed exceptions.  <p>An exception may have suppressed exceptions while also being caused by another exception.  Whether or not an exception has a cause is semantically known at the time of its creation, unlike whether or not an exception will suppress other exceptions which is typically only determined after an exception is thrown.  <p>Note that programmer written code is also able to take advantage of calling this method in situations where there are multiple sibling exceptions and only one can be propagated.</td>
<td align="left">NullPointerException<br/>IllegalArgumentException</td>
</tr>
<tr>
<td align="center"><strong>getSuppressed(): Throwable[]</strong></td>
<td align="center">public</td>
<td align="center">否</td>
<td align="center"><strong><font color='red'>是</font></strong></td>
<td align="center">否</td>
<td align="center"><strong><font color='red'>是</font></strong></td>
<td align="center">否</td>
<td align="center">否</td>
<td align="left">Returns an array containing all of the exceptions that were suppressed, typically by the ‘try’-with-resources statement, in order to deliver this exception.  If no exceptions were suppressed or ‘#Throwable(String, Throwable, boolean, boolean) suppression is disabled’, an empty array is returned.  This method is thread-safe.  Writes to the returned array do not affect future calls to this method.</td>
<td align="left"></td>
</tr>
</tbody></table>
<h4 id="内部类"><a href="#内部类" class="headerlink" title="内部类"></a>内部类</h4><h3 id="Throwable-SentinelHolder"><a href="#Throwable-SentinelHolder" class="headerlink" title="Throwable$SentinelHolder"></a><a href="./Throwable$SentinelHolder.html">Throwable$SentinelHolder</a></h3><h3 id="Throwable-PrintStreamOrWriter"><a href="#Throwable-PrintStreamOrWriter" class="headerlink" title="Throwable$PrintStreamOrWriter"></a><a href="./Throwable$PrintStreamOrWriter.html">Throwable$PrintStreamOrWriter</a></h3><h3 id="Throwable-WrappedPrintStream"><a href="#Throwable-WrappedPrintStream" class="headerlink" title="Throwable$WrappedPrintStream"></a><a href="./Throwable$WrappedPrintStream.html">Throwable$WrappedPrintStream</a></h3><h3 id="Throwable-WrappedPrintWriter"><a href="#Throwable-WrappedPrintWriter" class="headerlink" title="Throwable$WrappedPrintWriter"></a><a href="./Throwable$WrappedPrintWriter.html">Throwable$WrappedPrintWriter</a></h3>
        <div id="react-support-footer-root"></div>
      </div>
    </div>
  </div>
</div>

    


    

    <!-- js vendors -->
    <script src="//code.jquery.com/jquery-3.2.1.min.js" crossorigin="anonymous"></script>
    <script src="//cdnjs.cloudflare.com/ajax/libs/lunr.js/2.1.0/lunr.min.js"></script>

    <!-- js source  -->
    <script src="/java-source-code-analysis/script/doc.js"></script>

    

  </body>
</html>
